examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py

```diff
@@ -0,0 +1,690 @@
+                    core_context=core_context, latest_checkpoint=latest_checkpoint
+                )
+
+            # There will be a single op of len max_length, as defined in the searcher config.
+            for op in core_context.searcher.operations():
+                while trainer.steps_completed < op.length:
+                    for batch in trainer.train_dataloader:
+                        # Use the accumulate method for efficient gradient accumulation.
+                        with trainer.accelerator.accumulate(trainer.text_encoder):
+                            trainer._train_one_batch(batch)
+                        took_sgd_step = trainer.accelerator.sync_gradients
+                        if took_sgd_step:
+                            trainer.steps_completed += 1
+                            trainer.logger.info(f"Step {trainer.steps_completed} completed.")
+
+                            is_end_of_training = trainer.steps_completed == op.length
+                            time_to_report = (
+                                trainer.steps_completed % trainer.metric_report_freq == 0
+                            )
+                            time_to_ckpt = trainer.steps_completed % trainer.checkpoint_freq == 0
+
+                            # Report metrics, checkpoint, and preempt as appropriate.
+                            if is_end_of_training or time_to_report or time_to_ckpt:
+                                trainer._report_train_metrics(core_context)
+                                # report_progress for Web UI progress-bar rendering.
+                                if trainer.accelerator.is_main_process:
+                                    op.report_progress(trainer.steps_completed)
+                            if is_end_of_training or time_to_ckpt:
+                                trainer._save(core_context)
+                                if core_context.preempt.should_preempt():
+                                    return
+                            if is_end_of_training:
+                                break
+                if trainer.accelerator.is_main_process:
+                    # Report the final mean loss.
+                    op.report_completed(trainer.last_mean_loss)
+
+    def _train_one_batch(self, batch: TorchData) -> None:
+        """Train on a single batch and update internal metrics."""
+        self.text_encoder.train()
+        # Convert sample images to latent space.
+        prompts, img_tensors = batch
+        with torch.no_grad():
+            latent_dist = self.vae.encode(img_tensors).latent_dist
+            latents = latent_dist.sample()
+
+            # In 2112.10752, it was found that the latent space variance plays a large role in image
+            # quality.  The following scale factor helps to maintain unit latent variance.  See
+            # https://github.com/huggingface/diffusers/issues/437 for more details.
+            scale_factor = 0.18215
```

Comment:
praise: Good documentation of magic numbers

Reply:
Spent a while trying to track this one down. Very annoying.